server/pairing/challenge_management.go

```diff
@@ -0,0 +1,247 @@
+	"github.com/gorilla/sessions"
+	"go.uber.org/zap"
+)
+
+const (
+	// Session names
+	sessionChallenge = "challenge"
+	sessionBlocked   = "blocked"
+)
+
+type ChallengeError struct {
+	Text     string
+	HTTPCode int
+}
+
+func (ce *ChallengeError) Error() string {
+	return fmt.Sprintf("%s : %d", ce.Text, ce.HTTPCode)
+}
+
+func makeCookieStore() (*sessions.CookieStore, error) {
+	auth := make([]byte, 64)
+	_, err := rand.Read(auth)
+	if err != nil {
+		return nil, err
+	}
+
+	enc := make([]byte, 32)
+	_, err = rand.Read(enc)
+	if err != nil {
+		return nil, err
+	}
+
+	return sessions.NewCookieStore(auth, enc), nil
+}
+
+// ChallengeGiver is responsible for generating challenges and checking challenge responses
+type ChallengeGiver struct {
+	cookieStore *sessions.CookieStore
+	encryptor   *PayloadEncryptor
+	logger      *zap.Logger
+	authedIP    net.IP
+}
+
+func NewChallengeGiver(e *PayloadEncryptor, logger *zap.Logger) (*ChallengeGiver, error) {
+	cs, err := makeCookieStore()
+	if err != nil {
+		return nil, err
+	}
+
+	return &ChallengeGiver{
+		cookieStore: cs,
+		encryptor:   e.Renew(),
+		logger:      logger,
+	}, nil
+}
+
+func (cg *ChallengeGiver) getIP(r *http.Request) (net.IP, *ChallengeError) {
+	h, _, err := net.SplitHostPort(r.RemoteAddr)
+	if err != nil {
+		cg.logger.Error("getIP: h, _, err := net.SplitHostPort(r.RemoteAddr)", zap.Error(err), zap.String("r.RemoteAddr", r.RemoteAddr))
+		return nil, &ChallengeError{"error", http.StatusInternalServerError}
+	}
+	return net.ParseIP(h), nil
+}
+
+func (cg *ChallengeGiver) registerClientIP(r *http.Request) *ChallengeError {
+	hIP, err := cg.getIP(r)
+	if err != nil {
+		return err
+	}
+	cg.authedIP = hIP
+	return nil
+}
+
+func (cg *ChallengeGiver) validateClientIP(r *http.Request) *ChallengeError {
+	// If we haven't registered yet register the IP
+	if cg.authedIP == nil || len(cg.authedIP) == 0 {
+		err := cg.registerClientIP(r)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Then compare the current req RemoteIP with the authed IP
+	hIP, err := cg.getIP(r)
+	if err != nil {
+		return err
+	}
+
+	if !cg.authedIP.Equal(hIP) {
+		cg.logger.Error(
+			"request RemoteAddr does not match authedIP: expected '%s', received '%s'",
+			zap.String("expected", cg.authedIP.String()),
+			zap.String("received", hIP.String()),
+		)
+		return &ChallengeError{"forbidden", http.StatusForbidden}
+	}
+	return nil
+}
+
+func (cg *ChallengeGiver) getSession(r *http.Request) (*sessions.Session, *ChallengeError) {
+	s, err := cg.cookieStore.Get(r, sessionChallenge)
+	if err != nil {
+		cg.logger.Error("checkChallengeResponse: cg.cookieStore.Get(r, sessionChallenge)", zap.Error(err), zap.String("sessionChallenge", sessionChallenge))
+		return nil, &ChallengeError{"error", http.StatusInternalServerError}
+	}
+	return s, nil
+}
+
+func (cg *ChallengeGiver) generateNewChallenge(s *sessions.Session, w http.ResponseWriter, r *http.Request) ([]byte, *ChallengeError) {
+	challenge := make([]byte, 64)
```

Comment:
Ah, the two `64` values represent two different things.

The first one that generates the `auth` var is the higher value of the recommended length for a session authentication key:

> // It is recommended to use an authentication key with 32 or 64 bytes.

The second usage, the usage you mention here, can be any length at all. The challenge requires AES encryption using a 256 bit / 32 byte AES key, which is the upper value specified in the AES specification. The challenge length could be any number really as the security of the challenge response is anchored in the strength of AES encryption. I chose 64 bytes because 2^6 is more "square" than 2^5 (2^6 == 8^2), 2^4 is a little small and 2^8 is crazy too big (2^7 is big too but less square than 2^8). 2^8 is a nice number though, 2^8 == 4^4 == 16^2

Reply:
2^8 is prettier for sure! Thanks for the detailed explanation :)